File: 17-merge-sorted-lists/solution.go
// Merge Two Sorted Linked Lists
// Time: O(m + n), Space: O(1)

package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

type Solution struct{}

func (s *Solution) MergeTwoLists(l1, l2 *ListNode) *ListNode {
	dummy := &ListNode{Val: 0}
	curr := dummy
	
	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			curr.Next = l1
			l1 = l1.Next
		} else {
			curr.Next = l2
			l2 = l2.Next
		}
		curr = curr.Next
	}
	
	if l1 != nil {
		curr.Next = l1
	} else {
		curr.Next = l2
	}
	
	return dummy.Next
}

func main() {
	l1 := &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 4}}}
	l2 := &ListNode{Val: 1, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4}}}
	
	sol := &Solution{}
	merged := sol.MergeTwoLists(l1, l2)
	
	for merged != nil {
		if merged.Next != nil {
			fmt.Printf("%d ", merged.Val)
		} else {
			fmt.Printf("%d\n", merged.Val)
		}
		merged = merged.Next
	}
}

File: 17-merge-sorted-lists/solution.cpp
// Merge Two Sorted Linked Lists
// Time: O(m + n), Space: O(1)

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(0);
        ListNode* curr = dummy;
        
        while (l1 && l2) {
            if (l1->val <= l2->val) {
                curr->next = l1;
                l1 = l1->next;
            } else {
                curr->next = l2;
                l2 = l2->next;
            }
            curr = curr->next;
        }
        
        curr->next = l1 ? l1 : l2;
        
        return dummy->next;
    }
};

int main() {
    ListNode* l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
    ListNode* l2 = new ListNode(1, new ListNode(3, new ListNode(4)));
    
    Solution sol;
    ListNode* merged = sol.mergeTwoLists(l1, l2);
    
    while (merged) {
        cout << merged->val << (merged->next ? " " : "\n");
        merged = merged->next;
    }
    return 0;
}
